File: CrackingTheCodingInterview/Udemy/Monarchy.js
class Monarchy {
    //n-ary tree
    //represent the children as an array
    //children: [Jane, Farah]...
    //DFS, preorder 
    //instantiate an internal hashmap
    constructor(king) {
        this.king = new Person(king);
        this.personMap = new Map();
        this.personMap.set(this.king.name, this.king);
    }

    birth = function(child, parent) { //both child and parent are of people classes
        let childPerson = new Person(child);
        let parentPerson = this.personMap.get(parent);
        this.personMap.set(parent, parentPerson.push(childPerson))
        this.personMap.set(child, childPerson)
    }

    death = function(name) {
        let deadPerson = this.personMap.get(name);
        if (!deadPerson) {
            return null;
        }
        deadPerson.isAlive = false;
    }

    getOrderOfSuccession = function() {
        //traverse Tree here performs our pre order DFS
        const order = [];

        this._dfs(this.king, order);
        return order;
    }

    _dfs = function(currentPerson, order) {
        if (currentPerson.isAlive) {
            order.push(currentPerson.name);
        }
        for (let i = 0; i < currentPerson.children.length; i++) {
            _dfs(currentPerson.children[i], order)
        }
    }
}

class Person {
    constructor(name) {
        this.name = name;
        this.isAlive = true; //we'll just add a boolean and switch it to False
        this.children = [];
    }
}

File: CrackingTheCodingInterview/Udemy/Monarchy.java
package CrackingTheCodingInterview.Udemy;

import java.util.List;

public interface Monarchy {
    void birth(String child, String parent);
    void death(String name);
    List<String> getOrderOfSuccession();
}

// Monarchy mon = new Monarchy('Jake')
//birth('Catherine', 'Jake')
//birth('Tom', 'Jake')
//birth('Celine', 'Jake') --> birth orders
//birth('Mark', 'Catherine')
//the order of succession is Jake -> Catherine -> Farah -> Mark
//order of succession can be changed by a death in the family
//data structure to retain a tree
//use any features in your language
//N-ary tree, can use DFS & BFS